src/_P093_MitsubishiAC.ino

```diff
@@ -0,0 +1,805 @@
+    }
+    return String(F("<unknown> ")) + state;
+  }
+
+  static void dumpPacket(const uint8_t* packet, size_t length, String& result) {
+    for (size_t idx = 0; idx < length; ++idx) {
+      result += formatToHex(packet[idx], F(""));
+      result += ' ';
+    }
+  }
+
+  static String dumpOutgoingPacket(const uint8_t* packet, size_t length) {
+    String message = F("M-AC - OUT: ");
+    dumpPacket(packet, length, message);
+    return message;
+  }
+
+  static String dumpIncomingPacket(const uint8_t* packet, int length) {
+    String message = F("M-AC - IN: ");
+    dumpPacket(packet, length, message);
+    return message;
+  }
+  #endif
+
+private:
+  ESPeasySerial* _serial;
+  State _state;
+  boolean _fastBaudRate;
+  uint8_t _readBuffer[READ_BUFFER_LEN];
+  uint8_t _readPos;
+  unsigned long _writeTimeout;
+  Values _currentValues;
+  Values _wantedSettings;
+  uint8_t _infoModeIndex;
+  unsigned long _statusUpdateTimeout;
+  boolean _tempMode;
+  boolean _wideVaneAdj;
+  boolean _valuesInitialized;
+  WriteStatus _writeStatus;
+  const Mappings _mappings;
+};
+
+boolean Plugin_093(byte function, struct EventStruct *event, String& string) {
+  boolean success = false;
+
+  switch (function) {
+
+    case PLUGIN_DEVICE_ADD: {
+      Device[++deviceCount].Number = PLUGIN_ID_093;
+      Device[deviceCount].Type = DEVICE_TYPE_DUAL;
+      Device[deviceCount].VType = SENSOR_TYPE_STRING;
+      Device[deviceCount].ValueCount = 1;
+      Device[deviceCount].SendDataOption = true;
+      Device[deviceCount].TimerOption = true;
+      Device[deviceCount].TimerOptional = true;
+      break;
+    }
+
+    case PLUGIN_GET_DEVICENAME: {
+      string = F(PLUGIN_NAME_093);
+      break;
+    }
+
+    case PLUGIN_GET_DEVICEVALUENAMES: {
+      strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[0], PSTR(PLUGIN_VALUENAME1_093));
+      break;
+    }
+
+    case PLUGIN_GET_DEVICEGPIONAMES: {
+      serialHelper_getGpioNames(event);
+      break;
+    }
+
+    case PLUGIN_WEBFORM_SHOW_CONFIG: {
+      string += serialHelper_getSerialTypeLabel(event);
+      success = true;
+      break;
+    }
+
+    case PLUGIN_WEBFORM_LOAD: {
+      serialHelper_webformLoad(event);
+      success = true;
+      break;
+    }
+
+    case PLUGIN_WEBFORM_SAVE: {
+      serialHelper_webformSave(event);
+      success = true;
+      break;
+    }
+
+    case PLUGIN_INIT: {
+      initPluginTaskData(event->TaskIndex, new P093_data_struct(CONFIG_PIN1, CONFIG_PIN2));
+      success = getPluginTaskData(event->TaskIndex) != nullptr;
+      break;
+    }
+
+    case PLUGIN_READ: {
+      P093_data_struct* heatPump = static_cast<P093_data_struct*>(getPluginTaskData(event->TaskIndex));
+      if (heatPump != nullptr) {
+        success = heatPump->read(event->String2);
+      }
+      break;
+    }
+
+    case PLUGIN_WRITE: {
+      P093_data_struct* heatPump = static_cast<P093_data_struct*>(getPluginTaskData(event->TaskIndex));
```

Comment:
Just looked here to see if your plugin also has support for commands.
But here you just echo any command given to the heatpump and if it returns true, then that command will not be tried on other plugins or controllers in ESPEasy.
So this `PLUGIN_WRITE` case should check to see if the first argument of the given command matches a command that the plugin can handle and only if it can handle the command return `success = true;`

Reply:
Currently `success = true` will be set if plugin actually handled the command. Is that not correct?

Reply:
Well you don't have a command prefix like `MitsubishiAC,....`
So if there is somehow a string that matches another plugin, then we may run into trouble.

Also if a plugin supports multiple instances (not for this plugin), it may be useful to have one parameter of the command reserved for the task index.

Reply:
Makes sense, will update ...

Reply:
Updated